### ISSUE TITLE
Postgres dialect fails to parse some string literals with embedded backslashes, resulting in message "Zero length Lex item returned from '<unlexable>'"

### ISSUE THREAD

**Issue by mrigger** [sqlfluff/sqlfluff#1069]
Consider the following test case:

```
$ cat test.sql
SELECT '\';
```
When executing `sqlfluff fix test.sql` (or `sqlfluff fix --dialect postgres test.sql`) the following output is printed:

```
==== finding fixable violations ====
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug. 
==== no fixable linting violations found ====
```
I think the warning is unexpected, since, for example, the statement is a valid Postgres or SQLite query.

I'm using sqlfluff version 0.5.6 and found this bug by trying it on some statements generated by [SQLancer](https://github.com/sqlancer/sqlancer). I assume it reproduces on the latest version, since executing the query using the online demo version results in a `Internal Server Error`.

**Comment by barrywhart** [comment]
IIUC, the backslash is being interpreted as an "escape" character for the following single quote, thus the quoted string is missing its closing quote.

**Comment by barrywhart** [comment]
I confirm that SQLite and Postgres are okay with this query. I think we need to update the string lexing regex for the Postgres dialect.

See the "String Constants With C-Style Escapes" section in the [Postgres docs](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS) for details. IIUC, Postgres is a bit smarter than some versions of SQL, and if the backslash is followed by something other than an expected value, it's interpreted as a normal character rather than an escape character.

**Comment by barrywhart** [comment]
Great catch, by the way!!

**Comment by mrigger** [comment]
I found what looks like a related bug or missing feature. The query below (which is valid in MySQL, Postgres, SQLite etc.) results in the same error `Zero length Lex item returned from '<unlexable>'. Report this as a bug.`
```sql
SELECT 0 | 0;
``` 
Detailed output:
```
$ sqlfluff lint test.sql
WARNING    Zero length Lex item returned from '<unlexable>'. Report this as a bug. 
== [test.sql] FAIL
L:   0 | P:   0 |  LXR | Fatal. Unable to lex characters: 0
```

**Comment by barrywhart** [comment]
I did some tinkering with regex, and this may be a starting point for a fix:
```
'(\\[bfnrt]|\\\d{1,3}|\\x[0-9A-Fa-f]{1,2})*'
```

However, this regex doesn't handle embedded single quotes, e.g.
```
select '''';
```

should return a string literal `'`.

**Comment by barrywhart** [comment]
The query `select 0 | 0` fails to parse for a different reason -- 0.5.6 does not support bitwise operators. There's an existing issue for this, #1051, which was addressed in a PR from a day or so ago, thus it'll be fixed in the next release.
